heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```diff
@@ -139,107 +142,129 @@ public HeronMasterDriver(EvaluatorRequestor requestor,
+      ContainerPlan containerPlan = packing.containers.get(id);
+      if (containerPlan == null) {
+        LOG.log(Level.WARNING, "There is no container for {0} in packing plan.", id);
+        return false;
       }
-      ActiveContext context = runningTaskToContextMap.remove(runningTask);
-      LOG.log(Level.INFO, "Killing task: {0}", runningTask.getId());
-      runningTask.close();
-      submitHeronExecutorTask(context);
+      Resource resource = containerPlan.resource;
+      worker = new HeronWorker(id, getCpuForExecutor(resource), getMemInMBForExecutor(resource));
+    } else {
+      AllocatedEvaluator evaluator = multiKeyWorkerMap.detachEvaluatorAndRemove(worker);
+      LOG.log(Level.INFO, "Shutting down container {0}", evaluator.getId());
+      evaluator.close();
     }
+
+    AllocatedEvaluator container = launchContainerForExecutor(worker);
+    return container != null;
   }
 
-  private void launchContainerForExecutor(final String executorId, final int cpu, final int mem)
-      throws InterruptedException {
-    executor.submit(new Callable<AllocatedEvaluator>() {
+  @VisibleForTesting
+  AllocatedEvaluator launchContainerForExecutor(final String executorId,
+                                                        final int cpu,
+                                                        final int mem) {
+    return launchContainerForExecutor(new HeronWorker(executorId, cpu, mem));
+  }
+
+  private AllocatedEvaluator launchContainerForExecutor(final HeronWorker worker) {
+    Future<AllocatedEvaluator> result = executor.submit(new Callable<AllocatedEvaluator>() {
       @Override
       public AllocatedEvaluator call() throws Exception {
-        AllocatedEvaluator evaluator = allocateContainer(executorId, cpu, mem);
+        AllocatedEvaluator evaluator = allocateContainer(worker.workerId, worker.cores, worker.mem);
+        if (evaluator == null) {
+          LOG.log(Level.WARNING, "Could not allocate container for id {0}", worker.workerId);
+          return null;
+        }
 
-        Configuration context = createContextConfig(executorId);
+        Configuration context = createContextConfig(worker.workerId);
 
         LOG.log(Level.INFO,
             "Activating container {0} for heron executor, id: {1}",
-            new Object[]{evaluator.getId(), executorId});
+            new Object[]{evaluator.getId(), worker.workerId});
         evaluator.submitContext(context);
-        reefContainerToHeronExecutorMap.put(evaluator.getId(), executorId);
+        multiKeyWorkerMap.assignEvaluatorToWorker(worker, evaluator);
         return evaluator;
       }
     });
+
+    try {
+      return result.get();
+    } catch (InterruptedException e) {
+      LOG.log(Level.SEVERE, "Interrupted while waiting for container", e);
+    } catch (ExecutionException e) {
+      LOG.log(Level.SEVERE, "Error while waiting for container", e);
+    }
+    return null;
```

Comment:
My dislike of nulls (one of them) is that they're so ambiguous in meaning. It's much more expressive to throw something like a a ContainerAllocationException, which could be extended with stats codes, reasons, response messages, call stacks, etc. Returning a null basically swallows all contextual information of what happened below, leaving the caller to just log an ambiguous message and move on.